jetty-ee9/jetty-ee9-webapp/src/test/java/org/eclipse/jetty/ee9/webapp/TempDirTest.java

```diff
@@ -140,42 +149,98 @@ public void jettyBaseWorkExists(WorkDir workDirExt) throws Exception
-        }
 
         //Let jetty create the tmp dir on the fly
-        Server server = new Server();
+        _server = new Server();
         WebAppContext webAppContext = new WebAppContext();
         webAppContext.setContextPath("/");
-        webAppContext.setWarResource(webAppContext.getResourceFactory().newResource(warFile));
-        server.setHandler(webAppContext);
-        server.start();
+        webAppContext.setWar(testWebappDir.toFile().getAbsolutePath());
+        _server.setHandler(webAppContext);
+        _server.start();
         File tempDirectory = webAppContext.getTempDirectory();
-        server.stop();
+        assertThat(webAppContext.getAttribute(ServletContext.TEMPDIR), is(webAppContext.getTempDirectory()));
+        _server.stop();
         assertNull(webAppContext.getTempDirectory());
         assertThat("Temp dir exists", !Files.exists(tempDirectory.toPath()));
     }
+
+    @Test
+    public void testExplicitTempDir(WorkDir workDir) throws Exception
+    {
+        Path myTempDir = workDir.getEmptyPathDir().resolve("my-temp-dir");
+        FS.ensureDirExists(myTempDir);
+
+        // Create war on the fly
+        Path testWebappDir = MavenTestingUtils.getProjectDirPath("src/test/webapp");
+
+        //Tell jetty what the temp dir is for the webapp
+        _server = new Server();
+        WebAppContext webAppContext = new WebAppContext();
+        webAppContext.setContextPath("/");
+        webAppContext.setWar(testWebappDir.toFile().getAbsolutePath());
+        webAppContext.setTempDirectory(myTempDir.toFile());
+        _server.setHandler(webAppContext);
+        _server.start();
+        File tempDirectory = webAppContext.getTempDirectory();
+        assertThat(webAppContext.getAttribute(ServletContext.TEMPDIR), is(tempDirectory));
+        assertThat(tempDirectory.toPath(), is(myTempDir));
+    }
+
+    @Test
+    public void testFreshTempDir(WorkDir workDir) throws Exception
+    {
+        // Create war on the fly
+        Path testWebappDir = MavenTestingUtils.getProjectDirPath("src/test/webapp");
+
+        //Test that if jetty is creating a tmp dir for the webapp, it is different on
+        //restart
+        _server = new Server();
+        WebAppContext webAppContext = new WebAppContext();
+        webAppContext.setContextPath("/");
+        webAppContext.setWar(testWebappDir.toFile().getAbsolutePath());
+        _server.setHandler(webAppContext);
+        _server.start();
+        File tempDirectory = webAppContext.getTempDirectory();
+        webAppContext.stop();
+        assertNull(webAppContext.getTempDirectory());
+        webAppContext.start();
+        assertThat(tempDirectory.toPath(), not(PathMatchers.isSame(webAppContext.getTempDirectory().toPath())));
+    }
+
+    @Test
+    public void testSameTempDir(WorkDir workDir) throws Exception
+    {
+        // Create war on the fly
+        Path testWebappDir = MavenTestingUtils.getProjectDirPath("src/test/webapp");
```

Comment:
same comment
```suggestion
        Path testWebappDir = MavenTestingUtils.getTargetPath("test-classes/webapp");
```

Reply:
I'm confused, the tests appear to be working on ee8?